### PR TITLE
Make searches paginated

### PR DESCRIPTION
**Pull request by 414owen**
Previously, searches only returned the first thirty results, and there was no way to access page two.

Note that this is a breaking API change.

### CODE REVIEW

**Comment by andreasabel** [comment]
Excellent!
There should be something added to the testsuite using the new API.  Can you look into adding tests, @414owen?

**Comment by andreasabel** [comment]
@mergifyio rebase

**Review by andreasabel — CHANGES_REQUESTED** [review]
Please supply a testcase or a demo in `samples/`.

**Comment by 414owen** [comment]
@andreasabel I've fixed up the existing samples, will this do?

**Review by andreasabel — CHANGES_REQUESTED** [review]
Good job!
Could you please add the samples to the executables in `samples/github-samples.cabal` so that they get built by the CI.
Concerning the tests you added, unfortunately they do not run at the moment on CI, see https://github.com/haskell-github/github/runs/6098329674?check_suite_focus=true#step:20:144 and issue 
- #478.

**Comment by 414owen** [comment]
Trying to use the `github-samples.cabal` project locally errors with `Could not find module ‘Github’`, but I presume CI will have more luck.

**Comment by 414owen** [comment]
@andreasabel Looks like code search is broken in `master`, because the `repository` field didn't line up 1:1 with the `Repo` datatype.

I've added a new type called CodeSearchRepo, and all the search samples have been tested locally.

**Comment by andreasabel** [comment]
Weird error message  by GHC 8 and 9 in CI failure (https://github.com/haskell-github/github/runs/6228048459?check_suite_focus=true#step:18:153):
```
Building executable 'github-search-repos' for github-samples-0..
[1 of 1] Compiling Main             ( Search/SearchRepos.hs, /__w/github/github/dist-newstyle/build/x86_64-linux/ghc-9.2.2/github-samples-0/x/github-search-repos/noopt/build/github-search-repos/github-search-repos-tmp/Main.o )

Search/SearchRepos.hs:43:34: error:
    Ambiguous occurrence ‘GitHub.repoName’
    It could refer to
       either the field ‘repoName’,
              imported qualified from ‘GitHub’ at Search/SearchRepos.hs:4:1-23
              (and originally defined in ‘GitHub.Data.Repos’)
           or the field ‘repoName’,
              imported qualified from ‘GitHub’ at Search/SearchRepos.hs:4:1-23
              (and originally defined in ‘GitHub.Data.Repos’)
   |
43 |   let fields = [ ("Name", show . GitHub.repoName)
   |                                  ^^^^^^^^^^^^^^^
```

GHC 7.8 / 7.10 fail with (https://github.com/haskell-github/github/runs/6228049019?check_suite_focus=true#step:18:46):
```
Building library for github-0.27.1..

src/GitHub/Data/Repos.hs:3:14:
    Unsupported extension: DuplicateRecordFields
```

File: src/GitHub/Data/Repos.hs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {-# LANGUAGE CPP               #-}
 {-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE DuplicateRecordFields #-}
```
Comment (andreasabel) [inline src/GitHub/Data/Repos.hs:3]: Maybe don't use the `DuplicateRecordFields` extension.  Are there strong reasons to use it?

Comment (414owen) [reply src/GitHub/Data/Repos.hs:3]: Not really, I used `DuplicateRecordFields` so I could copy-paste `Repo` fields into `CodeSearchRepo` fields. I'll remove it now.

**Comment by 414owen** [comment]
Ugh, GHC-7 semigroup issue -_-
One sec...

**Review by andreasabel — APPROVED** [review]

**Comment by andreasabel** [comment]
Thanks, @414owen , looks like everything is ready.  Squashing is ok? (Since there are fixup commits...)

**Comment by 414owen** [comment]
Sure thing!

**Pull request merged**